### PR TITLE
fix: Update cached theme dictionary when it should and add NumberBox updates back

### LINKED ISSUE #7123: Update NumberBox from WinUI sources
### Current behavior

https://github.com/unoplatform/uno/pull/7122

### Expected behavior

Revert back

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

_No response_

### Environment

_No response_

### NuGet package version(s)

_No response_

### Affected platforms

_No response_

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #7123 - Reverts #7122

## PR Type

What kind of change does this PR introduce?

- Bugfix

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 1]
```diff
@@ -48,7 +72,15 @@ public Uri Source
 		}
 
 		public IList<ResourceDictionary> MergedDictionaries => _mergedDictionaries;
-		public IDictionary<object, object> ThemeDictionaries => _themeDictionaries ??= new ResourceDictionary();
+		public IDictionary<object, object> ThemeDictionaries
+		{
+			get
+			{
+				_themeDictionaries ??= new ResourceDictionary();
+				RegisterOnKeyAdded();
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:80]: Registering on every get is probably expensive :)

I'd rather see an observable collection and a "version" integer or something similar to determine that the merged dictionary has changed. Also, can you make a different PR for this, specifically? Thanks!

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:80]: @jeromelaban My intention was to register only when it's re-assigned (ie, if it was null), but I forgot the `if` condition 😄 
Would that work or you think I should go for the an observable collection anyway?

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 2]
```diff
@@ -241,6 +277,10 @@ private void TryMaterializeLazy(in ResourceKey key, ref object value)
 				{
 					value = newValue;
 					_values[key] = newValue; // If Initializer threw an exception this will push null, to avoid running buggy initialization again and again (and avoid surfacing initializer to consumer code)
+					if (newValue is ResourceDictionary addedOrChangedResourceDictionary)
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:280]: Same here.

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs (hunk 1 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:80]: No, in retrospect it would not be enough. I'm wondering about the performance impact, though it'll affect setting values, not reading, so it's less of a problem.

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/NumberBox/NumberBox.Properties.cs [hunk 3]
```diff
@@ -40,16 +40,22 @@ public double Maximum
...
+		public static DependencyProperty MaximumProperty { get; } =
 			DependencyProperty.Register(nameof(Maximum), typeof(double), typeof(NumberBox), new FrameworkPropertyMetadata(double.MaxValue, (s, e) => (s as NumberBox)?.OnMaximumPropertyChanged(e)));
 
 		public double Value
 		{
 			get => (double)GetValue(ValueProperty);
-			set => SetValue(ValueProperty, value);
+			set
+			{
+				if (!Value.IsNaN() || !value.IsNaN())
```
Comment (lindexi) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/NumberBox/NumberBox.Properties.cs:51]: How about check IsNegative ?

Comment (Youssef1313) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/NumberBox/NumberBox.Properties.cs:51]: @lindexi This is ported from microsoft-ui-xaml repository.

**Comment by Youssef1313** [comment]
Will open another PR only for the caching issue per https://github.com/unoplatform/uno/pull/7125#discussion_r711043660
